x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/querydsl/query/TermsQuery.java

```diff
@@ -31,7 +34,16 @@ public TermsQuery(Source source, String term, List<Expression> values) {
         if (values.isEmpty()) {
             this.values = Collections.emptySet();
         } else {
-            this.values = new LinkedHashSet<>(Foldables.valuesOf(values, values.get(0).dataType()));
+            DataType dt = values.get(0).dataType();
+            Set<Object> set = new LinkedHashSet<>(CollectionUtils.mapSize(values.size()));
+            for (Expression e : values) {
+                if (e.foldable()) {
+                    set.add(DataTypeConverter.convert(e.fold(), dt));
+                } else {
+                    throw new QlIllegalArgumentException("Cannot determine value for {}", e);
```

Comment:
I guess this would be the case of something like:
```
SELECT * FROM test WHERE test.a = test.b
```

- Isn't this caught earlier?
  - Should we add a test for this exception?

Reply:
This is the intermediate query AST - in this case terms will be more like `A in (1, 2)`.
The check there is a translation of the existing code - which is defensive - if one of the values cannot be folded and it hasn't been caught, bail out.
An alternative would be to simply add the folded values directly.